pageserver/src/tenant/timeline/eviction_task.rs

```diff
@@ -194,15 +193,16 @@ impl Timeline {
         // NB: all the checks can be invalidated as soon as we release the layer map lock.
         // We don't want to hold the layer map lock during eviction.
         // So, we just need to deal with this.
-        let candidates: Vec<Arc<dyn PersistentLayer>> = {
+        let candidates: Vec<_> = {
             let guard = self.layers.read().await;
             let layers = guard.layer_map();
             let mut candidates = Vec::new();
             for hist_layer in layers.iter_historic_layers() {
                 let hist_layer = guard.get_from_desc(&hist_layer);
-                if hist_layer.is_remote_layer() {
-                    continue;
-                }
+
+                // funny: this is the best way to get local layers is to lock them into
+                // memory for the duration of eviction
+                let Ok(guard) = hist_layer.guard_against_eviction(false).await else { continue; };
```

Comment:
If there was a concurrent disk usage based eviction (dube) which had selected the same layers, but not yet called `evict_and_wait`, only effect would be that dube and eviction_task might both think they were the ones to evict this layer.

If this comes *after* `evict_and_wait` is called, it might succeed but most likely fail for eviction already having happened, and so only dube would think that it evicted this layer.

Reply:
> I don't understand why this function needs to guard the layer against eviction.

This is the only API I'd like to give out representing the state. Though, I guess with #5074 it's no longer an issue to provide others as well. Then again, less is more with APIs...

Reply:
https://github.com/neondatabase/neon/pull/4938#discussion_r1307301400

Reply:
> 
> 
> If there was a concurrent disk usage based eviction (dube) which had selected the same layers, but not yet called evict_and_wait, only effect would be that dube and eviction_task might both think they were the ones to evict this layer.
> 
> If this comes after evict_and_wait is called, it might succeed but most likely fail for eviction already having happened, and so only dube would think that it evicted this layer.
> 

Why not turn this into a comment above the 

```
                let guard = match hist_layer.keep_resident().await {
```

then?

Reply:
This is now ```let guard = match hist_layer.keep_resident().await { ... }`.

I don't really see a problem with this, but ... adding a comment.

```patch
+                // guard against eviction while we inspect it; it might be that eviction_task and
+                // disk_usage_eviction_task both select the same layers to be evicted, and
+                // seemingly free up double the space. both succeeding is of no consequence.
                 let guard = match hist_layer.keep_resident().await {
```

Reply:
f5a171076